pest/src/inputs_private/string_input.rs

```diff
@@ -0,0 +1,307 @@
+// pest. The Elegant Parser
+// Copyright (C) 2017  Dragoș Tiselice
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::ascii::AsciiExt;
+use std::ops::Range;
+use std::str;
+
+use super::Input;
+
+/// A `struct` useful for matching heap-allocated `String`s.
+#[derive(Debug)]
+pub struct StringInput {
+    string: String
+}
+
+impl StringInput {
+    /// Creates a new `StringInput` from a `String`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest::inputs::{Input, StringInput};
+    /// let input = StringInput::new("asd".to_owned());
+    ///
+    /// assert_eq!(input.len(), 3);
+    /// ```
+    pub fn new(string: String) -> StringInput {
+        StringInput {
+            string: string
+        }
+    }
+}
+
+impl Input for StringInput {
```

Comment:
I haven't documented this yet since I'm not so sure how to approach this. On one hand, I'd really like this to be as concise and short as possible and put this as a comment in the `inputs` `mod`. On the other hand, I'd like this to be really easy to figure out if they're hacking on pest and wondering what all these inlines are all about, so a comment documenting each one is also something that I would consider.